coderd/notifications/dispatch/smtp.go

```diff
@@ -133,47 +131,42 @@ func (s *SMTPHandler) dispatch(subject, htmlBody, plainBody, to string) Delivery
 			}
 		}()
 
-		// Server handshake.
-		hello, err := s.hello()
-		if err != nil {
-			return false, xerrors.Errorf("'hello' validation: %w", err)
-		}
-		err = c.Hello(hello)
-		if err != nil {
-			return false, xerrors.Errorf("server handshake: %w", err)
-		}
-
 		// Check for authentication capabilities.
-		// if ok, mech := c.Extension("AUTH"); ok {
-		//	auth, err := s.auth(mech)
-		//	if err != nil {
-		//		return true, xerrors.Errorf("find auth mechanism: %w", err)
-		//	}
-		//	if auth != nil {
-		//		if err := c.Auth(auth); err != nil {
-		//			return true, xerrors.Errorf("%T auth: %w", auth, err)
-		//		}
-		//	}
-		//}
+		if ok, avail := c.Extension("AUTH"); ok {
+			// Ensure the auth mechanisms available are ones we can use.
+			auth, err := s.auth(ctx, avail)
+			if err != nil {
+				return true, xerrors.Errorf("determine auth mechanism: %w", err)
+			}
+
+			// If so, use the auth mechanism to authenticate.
+			if auth != nil {
+				if err := c.Auth(auth); err != nil {
+					return true, xerrors.Errorf("%T auth: %w", auth, err)
+				}
+			}
+		} else if !s.cfg.Auth.Empty() {
+			return false, xerrors.New("no authentication mechanisms supported by server")
+		}
 
 		// Sender identification.
 		from, err := s.validateFromAddr(s.cfg.From.String())
 		if err != nil {
 			return false, xerrors.Errorf("'from' validation: %w", err)
 		}
-		err = c.Mail(from)
+		err = c.Mail(from, &smtp.MailOptions{})
```

Comment:
Do you think we can build a fluent interface? Something like:

`c.Mail().Rcpts(...).Data(...).Send()`

Reply:
We could consider it, but I don't personally see the value. I'm very willing to be convinced though.

Reply:
Readability, but I won't push for it. This syntax is also fine.